HTTP/2 103 Early Hints Support (Now reopened to PR #4114)

apu031 opened the pull request:
# Implement HTTP/2 103 Early Hints Support

Add complete HTTP/2 103 Early Hints implementation with full client and server support, building upon the initial 1xx informational responses prototype to deliver a solution for web performance optimization.

## Key Features Added

### Client-side Support

- Add `InformationalConfig` with callback system for processing 1xx informational responses
- Extend HTTP/2 client builder with `informational_responses()` configuration method  
- Implement informational response polling in h2 client task with callback invocation
- Support real-time processing of 103 Early Hints and other informational responses
- Handle multiple informational responses before final response arrives

### Server-side Support

- **Builder opt-in pattern**: Add `.enable_informational()` method for explicit feature activation
- **Zero-cost abstraction**: When disabled (default), zero overhead - no allocations, no extensions
- **early_hints_pusher() API**: Ergonomic interface for sending 103 Early Hints
  - Returns `EarlyHintsPusher` handle when feature enabled
  - Returns `EarlyHintsError::NotSupported` when feature not enabled
  - Validates 103 status code requirement
- Add `InformationalSender` extension for hint transmission via mpsc channel
- Implement server-side informational response forwarding using h2's send_informational API
- Support multiple 103 Early Hints responses per request for progressive resource discovery
- Validate informational responses and handle errors gracefully

### Complete Implementation

- Build upon Joel Dice's initial HTTP/1 prototype (commit 4e2483a1) to add full HTTP/2 support
- Integrate with h2 crate's informational response capabilities 
- Add robust error handling and validation
- Include production-ready example with TLS, resource preloading, and performance monitoring
- Provide extensive test coverage with 11 integration tests covering different scenarios

## Changed Files

- **Cargo.toml**: Example build configuration
- **examples/README.md**: Document new HTTP/2 Early Hints example
- **examples/http2_early_hints.rs**: Working example with TLS, and resource preloading
- **src/client/conn/http2.rs**: Extend HTTP/2 client builder with informational response configuration
- **src/client/conn/informational.rs**: New module implementing client-side informational callback system  
- **src/client/conn/mod.rs**: Export new informational module
- **src/ext/informational_sender.rs**: New `early_hints_pusher()` API
- **src/ext/mod.rs**: Export new Early Hints API types
- **src/proto/h1/dispatch.rs**: Update HTTP/1 dispatch for informational compatibility
- **src/proto/h1/role.rs**: Extend HTTP/1 role handling for informational responses
- **src/proto/h2/client.rs**: HTTP/2 client informational response polling
- **src/proto/h2/server.rs**: Add `enable_informational` config field (defaults to false), conditional channel creation, helper method for polling logic
- **src/server/conn/http2.rs**: Add `.enable_informational()` builder method
- **tests/integration-early-hints.rs**: 11 integration tests


## Test Coverage

### Integration Tests (11 tests, ~0.27s)
All tests pass consistently, covering:
  - Basic preconnect hints functionality
  - Resource preloading (CSS, JS, fonts, images) 
  - Content Security Policy enforcement via Early Hints
  - Multiple sequential 103 responses
  - Mixed Link header types with HTTP/2 compression handling
  - Cross-origin redirect security behavior
  - Real-world web page optimization scenarios
  - Empty 103 response validation
  - Timing validation for performance benefits
  - Error response handling after hints sent

### Unit Tests (18 tests, ~0.24s)
- **API layer** (7 tests): Error handling, channel communication, status validation
- **Builder layer** (6 tests): Configuration, method chaining, flag propagation  
- **Config layer** (5 tests): Defaults, mutation, cloning, debug output

### Documentation Tests (2 tests)
- Examples for `early_hints_pusher()` usage
- Example for `.enable_informational()` builder method

**Total: 152 tests passing (31 new from this PR)**


## API Usage Example

### Server-side (Requires Opt-in)

```rust
use hyper::server::conn::http2;
use hyper::ext::early_hints_pusher;
use hyper::{Request, Response, StatusCode};

// 1. Enable feature via builder
let builder = http2::Builder::new(executor)
    .enable_informational();

// 2. Use in service handler
async fn handle(mut req: Request<Incoming>) -> Result<Response<String>, Error> {
    if let Ok(mut pusher) = early_hints_pusher(&mut req) {
        let hints = Response::builder()
            .status(StatusCode::EARLY_HINTS)
            .header("link", "</style.css>; rel=preload; as=style")
            .header("link", "</app.js>; rel=preload; as=script")
            .body(())
            .unwrap();
        
        let _ = pusher.send_hints(hints).await;
    }
    
    Ok(Response::new("<!DOCTYPE html>...".to_string()))
}
```


## Use Cases Supported

- **103 Early Hints**: Enable early resource discovery and preloading for web performance optimization
- **100 Continue**: Support request/response optimization for large payloads
- **Resource Preloading**: CSS, JavaScript, fonts, images, and fetch resources
- **CDN Preconnection**: Early connection establishment to external domains
- **Progressive Enhancement**: Multiple hint responses as server processing progresses

## Technical Implementation

- Builder opt-in pattern ensures zero overhead when feature not used (default state)
- Informational responses sent without END_STREAM flag to maintain connection
- Stream state management preserves HTTP/2 flow control during informational responses  
- Client callback system separates informational processing from main response flow
- Server extension system allows middleware and applications to send hints seamlessly
- Cross-origin security validated for redirect scenarios
- Backward compatibility maintained - existing code continues to work unchanged

## Standards Compliance

- **RFC 8297**: Complete HTTP Status Code 103 Early Hints implementation
- **RFC 7540**: HTTP/2 specification compliance for informational responses
- **HTTP/2 HPACK**: Proper header compression handling for multiple Link headers
- **Browser Security**: Cross-origin redirect protection and validation

## Migration Path

This implementation is fully backward compatible with opt-in behavior:

### No Changes Required

Existing applications continue to work unchanged with zero performance impact.

### To Enable Feature

Add single builder method call:
```
// Before: Feature automatically disabled
let builder = http2::Builder::new(executor);

// After: Single line to enable
let builder = http2::Builder::new(executor)
    .enable_informational();
```

Then use early_hints_pusher() in handlers:
```
if let Ok(mut pusher) = early_hints_pusher(&mut req) {
    pusher.send_hints(hints).await?;
}
```

No breaking changes. Pay only for what you use.


---

## Browser Tests

- Chrome:
<img width="2054" height="1187" alt="Screenshot 2025-12-09 at 5 27 54 AM" src="https://github.com/user-attachments/assets/23ce1f9c-e9de-4831-b47b-743d196b3563" />

- Firefox:
<img width="2054" height="1187" alt="Screenshot 2025-12-09 at 5 29 33 AM" src="https://github.com/user-attachments/assets/468faa4e-77c4-4073-859d-bc282efc79ad" />

---

The implementation provides a robust foundation for web performance optimization while maintaining the reliability and safety expected from the hyper HTTP library.

Resolves #3980 and #2426 for HTTP/2 103 Early Hints support and extends the informational responses capability for HTTP/2 protocols.



---
---

## Revision 2

### Addresses @JakubKoralewski 's feedback

This implementation addresses all concerns raised during code review:

1. Unconditional allocation → Builder opt-in with zero-cost default
2. API design consistency → Follows `informational_pusher` pattern
3. Performance impact → Zero overhead when disabled
4. Redundant feature gates → Removed unnecessary annotations
5. Dependency version → h2 = 0.4.13
6. Example quality → Only critical resources hinted
7. Polling optimization → Restricted to Service state only

dicej committed:
prototype support for 1xx informational responses

This is a prototype intended to spur discussion about what support for 1xx
informational responses should look like in a Hyper server.  The good news is
that it works great (for HTTP/1 only, so far).  The bad news is it's kind of
ugly.  Here's what I did:

- Add `ext::InformationalSender`, a type which wraps a
  `futures_channel::mspc::Sender<Response<()>>`.  This may be added as an
  extension to an inbound `Request` by the Hyper server, and the application
  and/or middleware may use it to send one or more informational responses
  before sending the real one.

- Add code to `proto::h1::dispatch` and friends to add such an extension to each
  inbound request and then poll the `Receiver` end along with the future
  representing the final response.  If the app never sends any informational
  responses, then everything proceeds as normal.  Otherwise, we send those
  responses as they become available until the final response is ready.

TODO items:
- [ ] Also support informational responses in the HTTP/2 server.
- [ ] Determine best way to handle when the app sends an informational response
  with a non-1xx status code.  Currently we just silently ignore it.
- [ ] Come up with a less hacky API?
- [ ] Add test coverage.

Signed-off-by: Joel Dice <joel.dice@fermyon.com>

Commit:
feat(http2): implement 103 Early Hints support

Add complete HTTP/2 103 Early Hints implementation with client and server support:

- Add InformationalSender extension for server-side hint transmission via mpsc channel
- Create InformationalCallback system for client-side informational response handling
- Extend HTTP/2 client builder with informational_responses() configuration method
- Implement informational response polling in h2 client task with callback invocation
- Add server-side informational response forwarding using h2's send_informational API
- Include extensive integration tests covering multiple scenarios and edge cases
- Add complete working example with TLS, resource preloading, and performance monitoring
- Update Cargo.toml with local h2 dependency and example build configuration

The implementation enables servers to send resource preload hints before final responses,
allowing browsers to start downloading critical resources early and improve page load
performance. Clients can register callbacks to process 103 Early Hints and other
informational responses.

Closes #3980, #2426

apu031 commented:
Hello @dicej , @pablocm, @vikanezrimaya, @nox, @seanmonstar!

My commit originally builds on @dicej's prototype commit https://github.com/hyperium/hyper/commit/4e2483a101e286d0c80ca077cad1e594bdeb52c8 in PR - https://github.com/hyperium/hyper/pull/3815. However, this code is obsolete and does not build anymore. Therefore, I have included both the commits in my PR. @seanmonstar @nox let me know if this would be the right way to move forward.

The code changes follow the [proposed public API](https://gist.github.com/vikanezrimaya/037101dc7b28de37ef03a47569213236#api-based-on-the-old-server-push-proposal) suggested from @vikanezrimaya. After merging this, I will release a PR for Axum wrapper.

Note - The tests will pass once we merge the h2 repo's PR - https://github.com/hyperium/h2/pull/865

Thank you very much once again for your time and support. 🙏🏼

JakubKoralewski commented on src/proto/h2/server.rs:
```diff
@@ -11,11 +11,20 @@ use h2::{Reason, RecvStream};
 use http::{Method, Request};
 use pin_project_lite::pin_project;
 
+#[cfg(feature = "server")]
```
i think src/proto/h2/mod.rs already makes sure this whole file is #[cfg(server)]

JakubKoralewski reviewed (commented):
Hi, I'm interested in 10x support so thought I'd contribute a review. Thanks for this work. I am not a contributor just hope to make it easier to review for someone in charge.

Commit:
refactor(http2): implement builder opt-in pattern for 103 Early Hints

Refactors the HTTP/2 103 Early Hints implementation to use an explicit
builder opt-in pattern, addressing code review feedback about unconditional
resource allocation.

## Changes

### Core Implementation
- **src/proto/h2/server.rs**: Added `enable_informational` config field
  (defaults to false) and conditional channel creation per request
- **src/server/conn/http2.rs**: Added `.enable_informational()` builder
  method for explicit feature opt-in
- **src/ext/informational_sender.rs**: New module with `early_hints_pusher()`
  API that returns error when feature is disabled
- **src/ext/mod.rs**: Exported new Early Hints API types

### Examples & Tests
- **examples/http2_early_hints.rs**: Updated to use `.enable_informational()`
- **tests/integration-early-hints.rs**: All 11 tests updated to use builder
  opt-in pattern

### Dependencies
- **Cargo.toml**: Updated h2 dependency comment for version requirement

## Benefits

### Zero-Cost Guarantee
- When disabled (default): No channel allocation, no extensions, zero overhead
- When enabled: Channel created per request with opt-in cost
- Follows hyper's "pay for what you use" philosophy

### API Design
- Explicit configuration required via builder method
- Clear error when feature not enabled (`EarlyHintsError::NotSupported`)
- Follows recommended informational_pusher pattern from design discussions

## Test Results
- ✅ All 11 integration tests passing
- ✅ All 100 library tests passing
- ✅ Live server demonstration successful (curl shows proper 103 responses)

## Addresses Review Feedback
- Resolves concern about unconditional channel allocation per request
- Implements builder-level opt-in as recommended
- Maintains zero overhead when feature not used
- Clear migration path for users (add single builder method call)

## Breaking Changes
None. Feature is opt-in and disabled by default, maintaining backward
compatibility.

Fixes: Addresses code review comments on PR #3991

liuraym5 commented:
I'm @apu031's teammate and I've had the chance to review this PR and it LGTM. I would love to see it move forward. Hope the maintainers can find some time to take a look! 🙏

apu031 commented:
Hi @seanmonstar and @nox, please let me know how we can move forward with this. I think my PR looks a bit longer because of so many line change but that's all in the integration test (`tests/integration-early-hints.rs` - 1234 lines) and the example (`examples/http2_early_hints.rs` - 569 lines). Without those, it only becomes a small PR of 800 lines.

If you prefer, I can break it up more. But I really want to get it to the finish line.

Thank you for taking a look!

apu031 commented:
Reopened in PR #4114 due to accidental force-push upon deletion of the working branch.
